// verdaccio/build/lib/cli/commands/info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfoCommand = void 0;

var _clipanion = require("clipanion");

var _envinfo = _interopRequireDefault(require("envinfo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class InfoCommand extends _clipanion.Command {
  async execute() {
    this.context.stdout.write('\nEnvironment Info:');
    const data = await _envinfo.default.run({
      System: ['OS', 'CPU'],
      Binaries: ['node', 'yarn', 'npm', 'pnpm'],
      Virtualization: ['Docker'],
      Browsers: ['Chrome', 'Edge', 'Firefox', 'Safari'],
      npmGlobalPackages: ['verdaccio']
    });
    this.context.stdout.write(data);
    process.exit(0);
  }

}

exports.InfoCommand = InfoCommand;

_defineProperty(InfoCommand, "paths", [[`--info`], [`-i`]]);
