// rewiremock/lib/plugins/_common.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PASS = exports.NO = exports.YES = exports.standardWipeCheck = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var YES = 'yes';
var NO = 'no';
var PASS = true;

var onetoone = function onetoone(a) {
    return a;
};
var pass = function pass() {
    return PASS;
};
var empty = function empty() {
    return null;
};

var createPlugin = function createPlugin(plugin) {
    var result = (0, _assign2.default)({
        init: empty,
        fileNameTransformer: onetoone,
        wipeCheck: pass,
        shouldMock: pass
    }, plugin);
    return result;
};

var standardWipeCheck = function standardWipeCheck(stubs, moduleName) {
    if ((0, _constants.getExtensions)().some(function (ext) {
        return stubs[moduleName + ext];
    })) {
        return YES;
    }
};

exports.standardWipeCheck = standardWipeCheck;
exports.YES = YES;
exports.NO = NO;
exports.PASS = PASS;
exports.default = createPlugin;
