// rewiremock/lib/plugins/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _childOnly = require('./childOnly');

var _childOnly2 = _interopRequireDefault(_childOnly);

var _nodejs = require('./nodejs');

var _nodejs2 = _interopRequireDefault(_nodejs);

var _protectNodeModules = require('./protectNodeModules');

var _protectNodeModules2 = _interopRequireDefault(_protectNodeModules);

var _relative = require('./relative');

var _relative2 = _interopRequireDefault(_relative);

var _webpackAlias = require('./webpack-alias');

var _webpackAlias2 = _interopRequireDefault(_webpackAlias);

var _toBeUsed = require('./toBeUsed');

var _toBeUsed2 = _interopRequireDefault(_toBeUsed);

var _directChild = require('./directChild');

var _directChild2 = _interopRequireDefault(_directChild);

var _disabledByDefault = require('./disabledByDefault');

var _disabledByDefault2 = _interopRequireDefault(_disabledByDefault);

var _mockThoughByDefault = require('./mockThoughByDefault');

var _mockThoughByDefault2 = _interopRequireDefault(_mockThoughByDefault);

var _usedByDefault = require('./usedByDefault');

var _usedByDefault2 = _interopRequireDefault(_usedByDefault);

var _toMatchOrigin = require('./toMatchOrigin');

var _toMatchOrigin2 = _interopRequireDefault(_toMatchOrigin);

var _nodeLibBrowser = require('./node-lib-browser');

var _nodeLibBrowser2 = _interopRequireDefault(_nodeLibBrowser);

var _mock__ = require('./__mock__');

var _mock__2 = _interopRequireDefault(_mock__);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mockThroughByDefault = _mockThoughByDefault2.default;

var webpack = _nodejs2.default;

var plugins = {
    childOnly: _childOnly2.default,
    nodejs: _nodejs2.default,
    webpack: webpack,
    protectNodeModules: _protectNodeModules2.default,
    relative: _relative2.default,
    webpackAlias: _webpackAlias2.default,
    toBeUsed: _toBeUsed2.default,

    disabledByDefault: _disabledByDefault2.default,

    mockThoughByDefault: _mockThoughByDefault2.default,
    mockThroughByDefault: mockThroughByDefault,

    alwaysMatchOrigin: _toMatchOrigin2.default,
    usedByDefault: _usedByDefault2.default,
    directChild: _directChild2.default,

    __mock__: _mock__2.default,

    nodeLibBrowser: _nodeLibBrowser2.default
};

exports.default = plugins;
