// rewiremock/es/plugins/index.js
import childOnly from './childOnly';
import nodejs from './nodejs';
import protectNodeModules from './protectNodeModules';
import relative from './relative';
import webpackAlias from './webpack-alias';
import toBeUsed from './toBeUsed';
import directChild from './directChild';

import disabledByDefault from './disabledByDefault';
import mockThoughByDefault from './mockThoughByDefault';
import usedByDefault from './usedByDefault';
import alwaysMatchOrigin from './toMatchOrigin';

import nodeLibBrowser from './node-lib-browser';
import __mock__ from './__mock__';

var mockThroughByDefault = mockThoughByDefault;

var webpack = nodejs;

var plugins = {
    childOnly: childOnly,
    nodejs: nodejs,
    webpack: webpack,
    protectNodeModules: protectNodeModules,
    relative: relative,
    webpackAlias: webpackAlias,
    toBeUsed: toBeUsed,

    disabledByDefault: disabledByDefault,

    mockThoughByDefault: mockThoughByDefault,
    mockThroughByDefault: mockThroughByDefault,

    alwaysMatchOrigin: alwaysMatchOrigin,
    usedByDefault: usedByDefault,
    directChild: directChild,

    __mock__: __mock__,

    nodeLibBrowser: nodeLibBrowser
};

export default plugins;
